dgram: add source-specific multicast support

LPardue opened the pull request:
This adds [RFC 4607](https://tools.ietf.org/html/rfc4607) support for IPv4 and IPv6. I'm opening this PR to start a discussion, it is not fully realised yet.

This changeset depends on [libuv PR 964](https://github.com/libuv/libuv/pull/964) and preempts that change being merged into node deps (i.e. it won't compile against master at the time of writing).

In approaching this change I considered modify the existing `addMembership/dropMembership` functions with a new sourceAddress parameter vs. adding `addSourceSpecificMembership/dropSourceSpecificMembership`. Both have merits but I picked the latter, I'm not stuck on it though. It comes does to opinion on which looks worse: `socket.addMembership(232.1.1.1, null, 123.45.67.89)` or `socket.addSourceSpecificMembership(123.45.67.89, 232.1.1.1)`.

I've tested this manually on a local network but don't currently have any tests defined. I'm not sure what node is looking for in that respect, testing source-specific multicast is hairy.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
dgram

jasnell commented on doc/api/dgram.md:
```diff
@@ -95,6 +95,21 @@ Tells the kernel to join a multicast group at the given `multicastAddress` and
 
+### socket.addSourceSpecificMembership(sourceAddress, groupAddress[, multicastInterface])
+<!-- YAML
+added: vX.Y.Z
```
nit: `added: REPLACEME`

LPardue replied:
fixed in 305693a7b5bea1fa86eee5ca12376cc9753b21b6

jasnell commented on lib/dgram.js:
```diff
@@ -222,7 +222,7 @@ Socket.prototype.bind = function(port_, address_ /*, callback*/) {
   // resolve address first
   this._handle.lookup(address, (err, ip) => {
-    if (err) {
+    if (err) {          
```
nit: extra whitespace added at the end here.

cjihrig replied:
You can have git remove this for you automatically. It looks like there is whitespace in other places too. If you take care of the problem now, we won't have commits removing unrelated whitespace in the future.  😄

LPardue replied:
Weird how that got in there sorry, fixed in 305693a.

jasnell commented:
Good to see this and thank you for the PR! First scan through looks good!

cjihrig commented on doc/api/dgram.md:
```diff
@@ -226,6 +241,24 @@ never have reason to call this.
 
+### socket.dropSourceSpecificMembership(sourceAddress, groupAddress[, multicastInterface])
+<!-- YAML
+added: vX.Y.Z
```
It's already pointed out above, but `added: REPLACEME` here as well. The reason is so that it will be properly detected and replaced during the release process.

LPardue replied:
fixed in 305693a

cjihrig commented on lib/dgram.js:
```diff
@@ -631,6 +631,49 @@ Socket.prototype.dropMembership = function(multicastAddress,
+                                                        interfaceAddress) {
+  this._healthCheck();
+
+  if (!sourceAddress) {
```
Can you make these argument checks stricter by checking for specific supported data types.

cjihrig replied:
I would drop the `if (!sourceAddress)` (and similar checks below) and just leave the check for `typeof === 'string'`. In general, I'd keep away from `if (var)` and `if (!var)` when doing type checking because they can match so many different things.

LPardue replied:
fixed

cjihrig commented on lib/dgram.js:
```diff
@@ -631,6 +631,49 @@ Socket.prototype.dropMembership = function(multicastAddress,
+    throw new errors.TypeError('ERR_MISSING_ARGS', 'groupAddress');
+  }
+
+  var err = this._handle.addSourceSpecificMembership(sourceAddress,
```
Please use `const` here.

cjihrig commented on lib/dgram.js:
```diff
@@ -631,6 +631,49 @@ Socket.prototype.dropMembership = function(multicastAddress,
+                                                         interfaceAddress) {
+  this._healthCheck();
+
+  if (!sourceAddress) {
```
Same comments apply to this function.

cjihrig commented on src/udp_wrap.cc:
```diff
@@ -324,6 +328,43 @@ void UDPWrap::DropMembership(const FunctionCallbackInfo<Value>& args) {
 }
 
+void UDPWrap::SetSourceMembership(const FunctionCallbackInfo<Value>& args,
+                            uv_membership membership) {
```
Please line `uv_membership` up vertically with `const` from the previous line.

LPardue replied:
fixed in 305693a

cjihrig commented on src/udp_wrap.cc:
```diff
@@ -324,6 +328,43 @@ void UDPWrap::DropMembership(const FunctionCallbackInfo<Value>& args) {
+  node::Utf8Value iface(args.GetIsolate(), args[2]);
+
+  const char* iface_cstr = *iface;
+  if (args[2]->IsUndefined() || args[2]->IsNull()) {
```
I'm not sure if we want to support `null` here, as that is generally intentional, where `undefined` is not. You can also move `node::Utf8Value iface(args.GetIsolate(), args[2]);` from above into the `else` branch so we don't compute it for no reason.

LPardue replied:
This bit of code is a bit tricky. I've tried to follow your suggestion, inverting the check and then assigning the pointer only if its a real thing.

As stated in my earlier comment, the function is based off of an existing one with a similar interface. Seemingly, it is permissable to explicitly call these functions with `null`, which will trigger a call down into libuv with `nullptr` effectively a no interface defined. 

I'm happy to change this but would really want the dgram API to be consistent across multicast functions. So any changes would probably need to be unilateral.

cjihrig commented on src/udp_wrap.cc:
```diff
@@ -324,6 +328,43 @@ void UDPWrap::DropMembership(const FunctionCallbackInfo<Value>& args) {
+  }
+
+  int err = uv_udp_set_source_membership(&wrap->handle_,
+                                  *groupAddress,
```
Please line up the arguments vertically.

LPardue replied:
fixed in 305693a

LPardue commented:
Thanks for the comments. I'll address them soon. One observation is that this is pretty much a copy/paste job based on the existing functions. I.e. the existing code probably also suffers from the issues raised. Would you like me to reflect back changes to the existing functions?

mcollina commented:
I'm a bit uncomfortable in landing this without tests, I know very well that testing multicast is complicated.

LPardue commented:
Thanks for the feesback. Are there any resources designing tests like this in node? How much of libuv tests (or methodology and test infrastructure) could be reused. This area is all bee to me but I completely understand the importance of testing before landing.

LPardue commented:
@cjihrig I've addressed your final review comment about type checking.

@mcollina I've added two tests for source-specific multicast (IPv4 and IPv6). These are near identical to the existing test `test-dgram-multicast-multi-process.js` test and so provide the same amount of test coverage. The main difference is that they discover a source address to use for the test., localhost doesn't work and will cause the receiver to fail.

cjihrig commented on test/internet/test-dgram-multicast-ssm-multi-process.js:
```diff
@@ -0,0 +1,251 @@
+// Copyright Joyent, Inc. and other Node contributors.
```
New files don't require this boilerplate at the top.

LPardue replied:
Shall I remove it completely?

cjihrig replied:
Yes, please.

cjihrig commented:
@LPardue  I went through the libuv PR. It needs updates from the author. I'd like another libuv collaborator to review it as well.

LPardue commented:
> @LPardue I went through the libuv PR. It needs updates from the author. I'd like another libuv collaborator to review it as well.

Indeed, the PR doesn't work on Windows without fixes I suggested there.

jasnell commented on test/internet/test-dgram-multicast-ssmv6-multi-process..js:
```diff
@@ -0,0 +1,230 @@
+
+    if (dead === listeners) {
+      console.error('[PARENT] All workers have died.');
+      console.error('[PARENT] Fail');
```
perhaps just use an `assert.fail()` here?

LPardue replied:
Think I've addressed this

jasnell commented on test/internet/test-dgram-multicast-ssmv6-multi-process..js:
```diff
@@ -0,0 +1,230 @@
+
+    killChildren(workers);
+
+    process.exit(1);
```
`assert.fail()` here also?

LPardue replied:
Think I've addressed this

addaleax commented on src/udp_wrap.cc:
```diff
@@ -324,6 +328,42 @@ void UDPWrap::DropMembership(const FunctionCallbackInfo<Value>& args) {
+
+  const char* iface_cstr = nullptr;
+  if (!args[2]->IsUndefined() && !args[2]->IsNull()) {
+    iface_cstr = *(node::Utf8Value(args.GetIsolate(), args[2]));
```
This is going to crash at some point – you need to keep the `Utf8Value` instance alive while `iface_cstr` is used. You’ll need to declare this `Utf8Value` like the others, then you can use `iface_cstr = *iface_utf8value`

LPardue replied:
@addaleax That's how it used to be coded but it was changed to address an earlier review comment by @cjihrig. 

> You can also move node::Utf8Value iface(args.GetIsolate(), args[2]); from above into the else branch so we don't compute it for no reason.

So I have conflicted advice and not sure how to resolve the issue.

cjihrig replied:
@LPardue my comment was because you were computing the variable, even when you didn't need it. In other words, `node::Utf8Value iface(args.GetIsolate(), args[2]);` was being executed no matter what, but the value was only used in one path.

I believe @addaleax is saying that you need to assign the value to a variable in the path where you do use it.

addaleax replied:
@cjihrig The thing is, this is really hard to write in a way that avoids the unnecessary computation for the undefined/null cases and is still somewhat readable… The best alternatives would probably be either using a `std::string` for memory management, or dynamically allocating the `Utf8Value` with `new`; both of these come with some runtime overhead as well.

LPardue replied:
@cjihrig I appreciate the issue but struggle to construct the appropriate node::Utf8Value in correct scope. The simple answer would be to branch and call `uv_udp_set_source_membership` in either branch (one with a populated `const char*` the other with `nullptr`). However, duplication is a code smell to me.

addaleax replied:
Unless this should be expected to be very hot code, I’d just go back to the previous state. (I am certainly not a domain expert, but it doesn’t sound like you’d ever want to call this method more than a handful of times in any Node process…)

cjihrig replied:
OK, yea, go with what you had before. Sorry for the noise.

LPardue replied:
My anticipated usage is to join SSM sessions at the start of a program and leave them at termination. In certain cases, join/leave requests may be made at a rate of up to 1 every 8 seconds. Other people's use cases may differ.

Thanks for the discussion, I'll revert back to original code.

addaleax replied:
Yeah, I think this is something that’s more on the order of 1 µs or so :smile:

LPardue replied:
@addaleax I believe I've addressed your review comment by implementing the change.

addaleax commented on src/udp_wrap.cc:
```diff
@@ -324,6 +328,42 @@ void UDPWrap::DropMembership(const FunctionCallbackInfo<Value>& args) {
+  CHECK_EQ(args.Length(), 3);
+
+  node::Utf8Value sourceAddress(args.GetIsolate(), args[0]);
+  node::Utf8Value groupAddress(args.GetIsolate(), args[1]);
```
Not that important but comon code style is `source_address`

LPardue replied:
Fixed

addaleax commented on doc/api/dgram.md:
```diff
@@ -226,6 +241,24 @@ never have reason to call this.
+
+* `sourceAddress` {string}
+* `groupAddress` {string}
+* `multicastInterface` {string}
```
Maybe the documentation should mention that this is optional (for both methods)?

LPardue replied:
I'm not terribly familiar with Node documentation. From what I've observed, the function signature indicates the optionality arguments by using square brackets e.g. `[arg]`. Note that the source-specific API is very close to the existing non-source-specific one. In both, the function description implies the optionality of `multicastInterface` rather than being fully explicit.

What are you suggesting for me to do?

jasnell commented on doc/api/dgram.md:
```diff
@@ -95,6 +95,21 @@ Tells the kernel to join a multicast group at the given `multicastAddress` and
+`IP_ADD_SOURCE_MEMBERSHIP` socket option. If the `multicastInterface` argument
+is not specified, the operating system will choose one interface and will add
+membership to it. To add membership to every available interface, call 
+`addSourceSpecificMembership` multiple times, once per interface.
```
minor style nit: s/addSourceSpecificMembership/addSourceSpecificMembership()`  (add the `()` to the end when mentioning functions)

LPardue replied:
Fixed, I emulated the style of the document to also reference the object e.g. `socket.addSourceSpecificMembership()`.

LPardue commented:
I think I've addressed review comments. The conflicting file relates to libuv and probably shouldn't be in the PR, I'll try to fix this soon.

This is really stalled on https://github.com/libuv/libuv/pull/964 which is awaiting changes in response to review.

refack commented:
Can one of the admins verify this patch?

jasnell commented:
Ping... what is blocking this?

LPardue commented:
Nothing's gonna happen on this until https://github.com/libuv/libuv/pull/964 lands.

Trott commented:
There was a merge conflict (in dgram.md), so I rebased against current master and resolved it. I also squashed the commits in this PR. IIUC, this is still stalled because we're waiting for https://github.com/libuv/libuv/pull/2202.

santigimeno commented:
`stalled` label removed as https://github.com/libuv/libuv/pull/2202 has landed and will be available in the next `v1.x` version.

LPardue commented:
Cool stuff.

Trott commented:
https://github.com/libuv/libuv/pull/2202 landed and is in libuv 1.32.0.

Node.js core was updated to libuv 1.32.0 in df0e2e3625d1f72e8b04e05ffa13997a51e9f0d7.

So I think this is now unblocked. Removing the `blocked` label.

Trott commented:
It's been a while, so this could benefit from a quick re-review or two. @jasnell @mcollina @addaleax @cjihrig @nodejs/dgram

Trott commented:
Getting a clean CI on this will probably require landing https://github.com/nodejs/node/pull/29979 first.

addaleax commented on src/udp_wrap.cc:
```diff
@@ -397,6 +401,43 @@ void UDPWrap::DropMembership(const FunctionCallbackInfo<Value>& args) {
+  node::Utf8Value group_address(args.GetIsolate(), args[1]);
+  node::Utf8Value iface(args.GetIsolate(), args[2]);
+
+  const char* iface_cstr = *iface;
```
I would also add `if (*iface == nullptr) return;`. We don’t have an explicit typecheck for `iface`, so converting it to string could throw an exception, and it would imo be odd to have this function throw an exception but still perform the call to `uv_udp_set_source_membership` in that case

addaleax commented:
@LPardue I’ve pushed a fix for my comment, included the commit you referenced, and a small fixup to make documentation linting pass (due to relatively new changes to the doc markdown format). I think after that this only requires us to get CI to pass.

n-thumann commented:
I noticed two deprecation warnings, caused by `this._healthCheck()`and `this._handle`. I fixed both of them in https://github.com/n-thumann/node/commit/d26529560e957b1756844a9ea4250e5e761219ff. Would you mind merging this commit too? :)

addaleax commented:
@n-thumann Done! (with linting issues fixed)

addaleax commented:
Landed in 1784b7fafac2755f870db3de9eb45a754b7a6477 :tada: Thanks for the PR, and sorry for all the delays! :sparkles:
